unit test for chroma_cens is failing except on py3.4

[Tagging @stefan-balke ]

I haven't investigated thoroughly, but I suspect that something is going wrong in the underlying BLAS system that can differ on travis by python version.  Either we need to relax the test criteria, or work out if there's a numerical instability that can be resolved within the cens calculation.

I just tested this out on my local conda environments:

- 2.7
  - numpy 0.11.3, scipy 0.18.1
  - test_cens passes
- 3.4
  - numpy 0.11.3, scipy 0.18.1
  - test_cens passes
- 3.5
  - numpy 0.12.1, scipy 0.19.0
  - test_cens fails

The culprit appears to be scipy: regressing my 3.5 environment back to scipy 0.18.1 fixes the problem.

I added a debug print to the test assertion, and it reads as follows:
```
======================================================================
FAIL: test_features.test_cens
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/bmcfee/miniconda/envs/py35/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/bmcfee/git/librosa/tests/test_features.py", line 644, in test_cens
    atol=1e-15), np.abs(ct_chroma_cens['f_CENS'] - lr_chroma_cens)
AssertionError: [[ 0.          0.          0.          0.          0.          0.          0.
   0.          0.          0.          0.          0.          0.          0.
   0.          0.          0.          0.          0.          0.          0.
   0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.          0.          0.
   0.          0.          0.          0.          0.          0.          0.
   0.          0.          0.          0.          0.          0.          0.
   0.          0.          0.        ]
 [ 0.          0.          0.          0.          0.00510423  0.05317791
   0.11701583  0.01799788  0.05074901  0.10170009  0.07319766  0.0084889
   0.04390135  0.05750489  0.03900799  0.02897911  0.10282193  0.05752425
   0.00286278  0.00159912  0.03659663  0.03217601  0.01264462  0.        ]
 [ 0.          0.          0.          0.          0.00127606  0.0074886
   0.00592027  0.00840787  0.00911527  0.          0.          0.00101242
   0.01000203  0.01432603  0.00554525  0.01845597  0.00380086  0.          0.
   0.          0.          0.          0.          0.        ]
 [ 0.          0.00093904  0.01237351  0.01335548  0.00709557  0.01512221
   0.00109966  0.00840787  0.00056326  0.03463202  0.01866429  0.03056786
   0.02067775  0.0082438   0.01133199  0.01035333  0.03713194  0.04624908
   0.00021418  0.0649565   0.04717409  0.00376252  0.          0.        ]
 [ 0.01025732  0.01859923  0.01846763  0.00743279  0.00487884  0.00731958
   0.01007773  0.00853884  0.01120566  0.00543333  0.01449387  0.01350839
   0.02425593  0.00113615  0.03386129  0.02206779  0.          0.00831296
   0.03160963  0.05040953  0.00703593  0.04345784  0.02160525  0.03512584]
 [ 0.00979378  0.0223595   0.02995336  0.00452205  0.          0.
   0.00428038  0.03280668  0.08255561  0.05268341  0.06619911  0.04334393
   0.03411163  0.02306906  0.06887878  0.05025624  0.05783308  0.01857221
   0.01668407  0.00668233  0.05104746  0.10352911  0.05661867  0.00536595]
 [ 0.00418841  0.00195403  0.04933958  0.13622121  0.08496527  0.00035907
   0.0077595   0.02199149  0.00380851  0.          0.          0.          0.
   0.          0.          0.          0.          0.          0.00307522
   0.04082566  0.0911065   0.06587543  0.01466293  0.01049256]
 [ 0.          0.          0.          0.          0.          0.          0.
   0.          0.          0.          0.          0.          0.          0.
   0.          0.          0.          0.          0.          0.          0.
   0.          0.          0.        ]
 [ 0.00074198  0.0211397   0.0947096   0.06624429  0.07614968  0.13074764
   0.05379601  0.          0.00440431  0.05075132  0.10244317  0.03049978
   0.06997875  0.10196474  0.05240886  0.00140105  0.02516837  0.04670977
   0.01895316  0.05231152  0.05641503  0.00752504  0.          0.        ]
 [ 0.          0.          0.00134425  0.01319038  0.02416936  0.00542621
   0.03134906  0.02622701  0.00380851  0.          0.          0.          0.
   0.          0.          0.          0.          0.          0.          0.
   0.          0.          0.          0.        ]
 [ 0.          0.          0.00537699  0.05276152  0.09795352  0.02919343
   0.11947598  0.11331593  0.02434929  0.          0.          0.          0.
   0.          0.          0.          0.          0.          0.          0.
   0.          0.          0.          0.        ]]

----------------------------------------------------------------------
Ran 1 test in 0.010s
```

As far as I can tell, the only relevant changes in scipy are in the `signal` module listed [here](https://scipy.github.io/devdocs/release.0.19.0.html#scipy-signal-improvements), specifically `convolve2d`.  I think it's now auto-switching to fftconvolve without telling us, and the default padding mode is killing us.

If I add `boundary='fill'` to the `convolve2d` call in `chroma_cens`, the test passes.

This makes me think: why do we need a 2d convolution here?  The filter is one-dimensional, and applied independently to each pitch class channel.  Should this not just be a 1d convolution?

Anyway, I'll PR the fix for this.

--- a/.travis.yml
+++ b/.travis.yml
@@ -29,6 +29,7 @@ before_install:
     - export PATH="$HOME/env/miniconda$TRAVIS_PYTHON_VERSION/bin:$PATH";
     - hash -r
     - source activate test-environment
+    - conda list
 
 install:
     # install your own package into the environment

--- a/librosa/feature/spectral.py
+++ b/librosa/feature/spectral.py
@@ -1097,7 +1097,7 @@ def chroma_cens(y=None, sr=22050, C=None, hop_length=512, fmin=None,
     win /= np.sum(win)
     win = np.atleast_2d(win)
 
-    cens = scipy.signal.convolve2d(chroma_quant, win, mode='same')
+    cens = scipy.signal.convolve2d(chroma_quant, win, mode='same', boundary='fill')
 
     # L2-Normalization
     return util.normalize(cens, norm=norm, axis=0)

--- a/librosa/feature/spectral.py
+++ b/librosa/feature/spectral.py
@@ -1093,11 +1093,12 @@ def chroma_cens(y=None, sr=22050, C=None, hop_length=512, fmin=None,
         chroma_quant += (chroma > cur_quant_step) * QUANT_WEIGHTS[cur_quant_step_idx]
 
     # Apply temporal smoothing
-    win = scipy.signal.hanning(win_len_smooth + 2, sym=False)
+    win = scipy.signal.hanning(win_len_smooth + 2)
     win /= np.sum(win)
     win = np.atleast_2d(win)
 
-    cens = scipy.signal.convolve2d(chroma_quant, win, mode='same', boundary='fill')
+    cens = scipy.signal.convolve2d(chroma_quant, win,
+                                   mode='same', boundary='fill')
 
     # L2-Normalization
     return util.normalize(cens, norm=norm, axis=0)

--- a/librosa/feature/spectral.py
+++ b/librosa/feature/spectral.py
@@ -1097,7 +1097,7 @@ def chroma_cens(y=None, sr=22050, C=None, hop_length=512, fmin=None,
     win /= np.sum(win)
     win = np.atleast_2d(win)
 
-    cens = scipy.signal.convolve2d(chroma_quant, win, mode='same', boundary='wrap')
+    cens = scipy.signal.convolve2d(chroma_quant, win, mode='same', boundary='fill')
 
     # L2-Normalization
     return util.normalize(cens, norm=norm, axis=0)

--- a/librosa/feature/spectral.py
+++ b/librosa/feature/spectral.py
@@ -1097,7 +1097,7 @@ def chroma_cens(y=None, sr=22050, C=None, hop_length=512, fmin=None,
     win /= np.sum(win)
     win = np.atleast_2d(win)
 
-    cens = scipy.signal.convolve2d(chroma_quant, win, mode='same', boundary='fill')
+    cens = scipy.signal.convolve2d(chroma_quant, win, mode='same', boundary='wrap')
 
     # L2-Normalization
     return util.normalize(cens, norm=norm, axis=0)

--- a/tests/test_features.py
+++ b/tests/test_features.py
@@ -640,4 +640,5 @@ def test_cens():
         # load CENS-41-1 features
         ct_chroma_cens = load(os.path.join('data', cur_fn_ct_chroma_cens))
 
-        assert np.allclose(ct_chroma_cens['f_CENS'], lr_chroma_cens, rtol=1e-15, atol=1e-15)
+        maxdev = np.max(np.abs(ct_chroma_cens['f_CENS'] - lr_chroma_cens))
+        assert np.allclose(ct_chroma_cens['f_CENS'], lr_chroma_cens, rtol=1e-15, atol=1e-15), maxdev

--- a/tests/test_features.py
+++ b/tests/test_features.py
@@ -640,5 +640,5 @@ def test_cens():
         # load CENS-41-1 features
         ct_chroma_cens = load(os.path.join('data', cur_fn_ct_chroma_cens))
 
-        maxdev = np.max(np.abs(ct_chroma_cens['f_CENS'] - lr_chroma_cens))
+        maxdev = np.abs(ct_chroma_cens['f_CENS'] - lr_chroma_cens)
         assert np.allclose(ct_chroma_cens['f_CENS'], lr_chroma_cens, rtol=1e-15, atol=1e-15), maxdev

